symengine/functions.cpp

```diff
@@ -2703,6 +2703,109 @@ RCP<const Basic> uppergamma(const RCP<const Basic> &s, const RCP<const Basic> &x
     return make_rcp<const UpperGamma>(s, x);
 }
 
+bool PolyGamma::is_canonical(const RCP<const Basic> &n, const RCP<const Basic> &x)
+{
+    if(is_a_Number(*x) and not (rcp_static_cast<const Number>(x))->is_positive()) {
+        return false;
+    }
+    if (eq(*n, *zero)) {
+        if(eq(*x, *one)) {
+        	return false;
+        }
+        if(is_a<Rational>(*x)) {
+            auto x_ = rcp_static_cast<const Rational>(x);
+            auto num = x_->i.get_num();
+            auto den = x_->i.get_den();
+            if(den == 2 and (num == 1 or num == 3)) {
+                return false;
+            }
+            if(den == 4 and num == 1) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+
+std::size_t PolyGamma::__hash__() const
+{
+    std::size_t seed = POLYGAMMA;
+    hash_combine<Basic>(seed, *n_);
+    hash_combine<Basic>(seed, *x_);
+    return seed;
+}
+
+bool PolyGamma::__eq__(const Basic &o) const
+{
+    if (is_a<PolyGamma>(o) and
+        eq(*x_, *(static_cast<const PolyGamma &>(o).x_)) and
+        eq(*n_, *(static_cast<const PolyGamma &>(o).n_)))
+        return true;
+    return false;
+}
+
+int PolyGamma::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<PolyGamma>(o))
+    const PolyGamma &pg = static_cast<const PolyGamma &>(o);
+    if (neq(*n_, *(pg.n_))) {
+        return n_->__cmp__(*(pg.n_));
+    }
+    else {
```

Comment:
Formatting according to https://github.com/symengine/symengine/blob/master/doc/style_guide.md#whitespace